File: python/complex_gift/driver.py
"""File-level driver: complex NIfTI in, complex component maps out.

`pipeline.run_complex_ica` works on in-memory `(T, V)` arrays and returns maps over the
masked voxels only. This module is the thin layer that makes the package usable on real
data: it loads GIFT's two-file complex volumes, flattens them, runs the pipeline, expands
the maps back to full volume shape, and writes them out as complex NIfTI pairs.
"""

from pathlib import Path

import nibabel as nib
import numpy as np

from .complex_io import REAL_IMAG, read_complex, write_complex
from .pipeline import run_complex_ica


def unmask(S, mask, dims):
    """Expand masked maps back to full volume shape.

    S: (N, V_masked) complex. mask: (V,) bool. dims: spatial shape, prod(dims) == V.
    Returns (N, *dims) complex, zero outside the mask.
    """
    S = np.asarray(S, dtype=np.complex128)
    mask = np.asarray(mask, dtype=bool).ravel()
    N = S.shape[0]
    V = mask.size
    if int(np.prod(dims)) != V:
        raise ValueError(f"dims {tuple(dims)} hold {int(np.prod(dims))} voxels, mask has {V}")

    full = np.zeros((N, V), dtype=np.complex128)
    full[:, mask] = S
    return full.reshape(N, *dims)


def _load_subject(pair, complex_type):
    """Load one subject's two-file complex volume as ((T, V) array, dims, affine)."""
    first, second = pair
    z = read_complex(first, second, complex_type=complex_type)     # (*dims, T)
    if z.ndim < 2:
        raise ValueError(f"expected a 4-D complex volume, got shape {z.shape}")
    dims, T = z.shape[:-1], z.shape[-1]
    affine = nib.load(str(first)).affine
    V = int(np.prod(dims))
    return z.reshape(V, T).T, dims, affine                          # (T, V)


def run_from_files(subject_pairs, n_components, out_dir, complex_type=REAL_IMAG,
                   prefix="component", naming=("R_", "I_"), **kwargs):
    """Run group complex ICA over subjects given as two-file complex NIfTI pairs.

    subject_pairs: list of (first_path, second_path). Each pair is one subject's 4-D
        complex volume (real&imaginary, or magnitude&phase per `complex_type`).
    n_components: model order.
    out_dir: directory for the written component maps (created if absent).
    kwargs: forwarded to `run_complex_ica` (e.g. `estimator`, `rng`, `n_subject`).

    Returns (result, written) where `result` is the PipelineResult and `written` is a list
    of (first_path, second_path) — one complex NIfTI pair per component.
    """
    if not subject_pairs:
        raise ValueError("no subjects given")

    loaded = [_load_subject(p, complex_type) for p in subject_pairs]
    dims = loaded[0][1]
    for _, d, _ in loaded[1:]:
        if d != dims:
            raise ValueError(f"subjects disagree on volume shape: {dims} vs {d}")
    affine = loaded[0][2]
    subject_data = [x for x, _, _ in loaded]

    result = run_complex_ica(subject_data, n_components=n_components, **kwargs)

    maps = unmask(result.S_group, result.mask, dims)                # (N, *dims)

    out_dir = Path(out_dir)
    out_dir.mkdir(parents=True, exist_ok=True)
    written = []
    for k, vol in enumerate(maps, start=1):
        base = f"{prefix}_{k:03d}.nii"
        first = out_dir / (naming[0] + base)
        second = out_dir / (naming[1] + base)
        write_complex(vol, first, second, affine=affine, complex_type=complex_type)
        written.append((first, second))

    return result, written

File: rust/src/driver.rs
//! File-level driver: complex NIfTI in, complex component maps out.
//!
//! `pipeline::run_complex_ica` is array-in/array-out and returns maps over the masked
//! voxels only. This is the thin layer that makes the crate usable on real data: it loads
//! GIFT's two-file complex volumes, flattens them, runs the pipeline, expands the maps back
//! to full volume shape, and writes them out as complex NIfTI pairs.
//!
//! Component NUMBERING IS NOT STABLE ACROSS PORTS OR RUNS. ICA fixes neither the order nor
//! the phase of its components, and no port canonicalizes them, so `component_001` here and
//! `component_001` from the Python driver are generally DIFFERENT sources - even for the
//! deterministic estimator, where the two agree to ~4e-14 once matched up. Match components
//! by correlation, never by filename.
//!
//! Mirrors `python/src/gift/driver.py`; the two ports are meant to be interchangeable.
//! Python's `unmask` returns an `(N, *dims)` array where this returns `N` flat volumes -
//! idiomatic per language, but the voxel ORDER is identical.

use std::path::{Path, PathBuf};

use nalgebra::DMatrix;
use num_complex::Complex64;

use crate::complex_io::{read_complex, write_complex_like, ComplexType};
use crate::pipeline::{run_complex_ica, Estimator, PipelineResult};

/// Expand masked maps back to full volume length, zero outside the mask.
///
/// `s` is `(N, V_masked)`; the returned `N` volumes each have `dims[0]*dims[1]*dims[2]`
/// entries, with column `c` of `s` scattered to the `c`-th set flat index of `mask`.
pub fn unmask(s: &DMatrix<Complex64>, mask: &[bool], dims: [usize; 3]) -> Vec<Vec<Complex64>> {
    let v = dims[0] * dims[1] * dims[2];
    assert_eq!(mask.len(), v, "mask length does not match dims");
    let kept: Vec<usize> = (0..v).filter(|&j| mask[j]).collect();
    assert_eq!(kept.len(), s.ncols(), "maps do not match the mask");

    (0..s.nrows())
        .map(|i| {
            let mut vol = vec![Complex64::new(0.0, 0.0); v];
            for (c, &j) in kept.iter().enumerate() {
                vol[j] = s[(i, c)];
            }
            vol
        })
        .collect()
}

/// Run group complex ICA over subjects given as two-file complex NIfTI pairs, and write
/// the group component maps back out as complex NIfTI pairs.
///
/// Returns the `PipelineResult` and one `(first, second)` path pair per component.
pub fn run_from_files(
    pairs: &[(PathBuf, PathBuf)],
    n_components: usize,
    out_dir: &Path,
    kind: ComplexType,
    estimator: Estimator,
    nf_dir: &Path,
) -> Result<(PipelineResult, Vec<(PathBuf, PathBuf)>), String> {
    if pairs.is_empty() {
        return Err("run_from_files: no subjects".into());
    }

    let mut subjects = Vec::with_capacity(pairs.len());
    let mut dims3: Option<[usize; 3]> = None;

    for (first, second) in pairs {
        let (data, d) = read_complex(first, second, kind)?;
        let this = [d[0], d[1], d[2]];
        match dims3 {
            None => dims3 = Some(this),
            Some(prev) if prev != this => {
                return Err(format!(
                    "subjects disagree on volume shape: {prev:?} vs {this:?}"
                ))
            }
            _ => {}
        }
        let v = this[0] * this[1] * this[2];
        let t = d[3];
        // `read_complex` returns the 4-D volume in C (standard) layout, so the LAST axis is
        // fastest: flat = ((x*ny + y)*nz + z)*nt + tt. TIME IS FASTEST, not slowest.
        // Indexing this as `data[tt * v + vv]` reads a scrambled volume with no error and
        // no shape change. Pinned by `reads_a_4d_volume_in_c_order_with_time_fastest` in
        // tests/complex_io.rs, and matches the Python port's `z.reshape(V, T).T`.
        let mut m = DMatrix::<Complex64>::zeros(t, v);
        for tt in 0..t {
            for vv in 0..v {
                m[(tt, vv)] = data[vv * t + tt];
            }
        }
        subjects.push(m);
    }
    let dims = dims3.expect("at least one subject");

    let res = run_complex_ica(&subjects, n_components, None, estimator, nf_dir)?;

    let vols = unmask(&res.s_group, &res.mask, dims);
    std::fs::create_dir_all(out_dir)
        .map_err(|e| format!("cannot create {}: {e}", out_dir.display()))?;

    let mut written = Vec::with_capacity(vols.len());
    for (k, vol) in vols.iter().enumerate() {
        let base = format!("component_{:03}.nii", k + 1);
        let first = out_dir.join(format!("R_{base}"));
        let second = out_dir.join(format!("I_{base}"));
        // Carry the first subject's geometry onto the maps, matching Python's
        // `write_complex(..., affine=affine)`. Without a reference the outputs get an
        // identity affine and will not overlay on the subject's anatomy.
        write_complex_like(vol, dims, &first, &second, kind, Some(&pairs[0].0))?;
        written.push((first, second));
    }

    Ok((res, written))
}
